application/src/main/java/org/togetherjava/tjbot/commands/filesharing/FileSharingMessageListener.java

```diff
@@ -0,0 +1,183 @@
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+
+/**
+ * Listener that receives all sent help messages and uploads them to a share service if the message
+ * contains a file with the given extension in the
+ * {@link FileSharingMessageListener#extensionFilter}
+ */
+public class FileSharingMessageListener extends MessageReceiverAdapter {
+
+    private static final ObjectMapper JSON = new ObjectMapper();
+
+    private final String gistApiKey;
+    private static final String SHARE_API = "https://api.github.com/gists";
+    private static final HttpClient CLIENT = HttpClient.newHttpClient();
+
+    private final Set<String> extensionFilter = Set.of("txt", "java", "gradle", "xml", "kt", "json",
+            "fxml", "css", "c", "h", "cpp", "py", "yml");
+
+    private final Predicate<String> isStagingChannelName;
+    private final Predicate<String> isOverviewChannelName;
+
+
+    public FileSharingMessageListener(@NotNull Config config) {
+        super(Pattern.compile(".*"));
+
+        gistApiKey = config.getGistApiKey();
+        isStagingChannelName = Pattern.compile(config.getHelpSystem().getStagingChannelPattern())
+            .asMatchPredicate();
+        isOverviewChannelName = Pattern.compile(config.getHelpSystem().getOverviewChannelPattern())
+            .asMatchPredicate();
+    }
+
+    @Override
+    public void onMessageReceived(@NotNull MessageReceivedEvent event) {
+        User author = event.getAuthor();
+        if (author.isBot() || event.isWebhookMessage()) {
+            return;
+        }
+
+        if (!isHelpThread(event)) {
+            return;
+        }
+
+        List<Message.Attachment> attachments =
+                event.getMessage().getAttachments().stream().filter(attachment -> {
+                    String extension = attachment.getFileExtension();
+                    if (extension == null) {
+                        return false;
+                    }
+                    return extensionFilter.contains(extension);
+                }).toList();
+
+        processAttachments(event, attachments);
+    }
+
+    private void processAttachments(@NotNull MessageReceivedEvent event,
+            @NotNull List<Message.Attachment> attachments) {
+        Map<String, GistFile> filesAsJson = new HashMap<>();
+
+        for (Message.Attachment attachment : attachments) {
+            attachment.retrieveInputStream()
+                .thenApply(this::readAttachment)
+                .thenAccept(content -> filesAsJson.put(createFileName(attachment),
+                        new GistFile(content)))
+                .join();
```

Comment:
Dont join on this directly. That way you compute it sequentially and not in-parallel. Huge bummers.
Also, this actually has a bug. You are modifying the map (which is not thread-safe) in a threaded environment.

Instead, stop at `thenAccept(...)`. The method gives you back a future, which you can then collect in a list of futures. Then, after launching all tasks, you can start waiting on them. So:

```java
Map<String, GistFile> nameToFile = Collections.synchronizedMap(new HashMap<>());

List<CompletableFuture<Void>> tasks = new ArrayList<>();
  CompletableFuture<Void> task = attachment.retrieveInputStream()
    .thenApply(this::readAttachment)
    .thenAccept(content -> filesAsJson.put(createFileName(attachment), new GistFile(content)));

  tasks.add(task);
}

tasks.forEach(CompletableFuture::join);
```

Reply:
That said, now you are actually running something blocking on a JDA thread (the thread of `onMessageReceived`), that is not acceptable and can block JDA itself if its very busy.

You have to make sure that those `join`s are actually executed on some other thread. The easiest way to achieve this is by posting the whole `proccessAttachments` method on a service:
```java
CompletableFuture.runAsync(() -> processAttachments(event, attachments));
```